File: frontend/src/types/agent.ts
export interface AgentContext {
  term_id?: number
  campus_id?: number
}

export interface AgentToolCall {
  name: string
  arguments: Record<string, unknown>
  success: boolean
  error?: string | null
}

export interface AgentChartSeries {
  name: string
  data: unknown[]
}

export interface AgentVisualization {
  type: 'bar' | 'line' | 'pie'
  title: string
  categories: string[]
  series: AgentChartSeries[]
}

export interface AgentChatResponse {
  conversation_id: number
  answer: string
  model_used: string
  tool_calls: AgentToolCall[]
  data: unknown
  visualization: AgentVisualization | null
}

export interface AgentConversation {
  id: number
  title: string
  message_count: number
  created_at: string
  updated_at: string
}

export interface AgentMessage {
  id: number
  role: 'user' | 'assistant'
  content: string
  model_used: string | null
  tool_calls: AgentToolCall[]
  created_at: string
}

export interface DictionaryItem {
  id: number
  name: string
}

File: backend/app/schemas/agent.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class StrictAgentModel(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class AgentContext(StrictAgentModel):
    term_id: int | None = Field(default=None, ge=1)
    campus_id: int | None = Field(default=None, ge=1)


class AgentChatRequest(StrictAgentModel):
    message: str = Field(min_length=1, max_length=2000)
    conversation_id: int | None = Field(default=None, ge=1)
    context: AgentContext = Field(default_factory=AgentContext)


class AgentToolCallView(BaseModel):
    name: str
    arguments: dict[str, Any]
    success: bool
    error: str | None = None


class AgentChartSeries(BaseModel):
    name: str
    data: list[Any]


class AgentVisualization(BaseModel):
    type: Literal["bar", "line", "pie"]
    title: str
    categories: list[str] = Field(default_factory=list)
    series: list[AgentChartSeries]


class AgentChatResponse(BaseModel):
    conversation_id: int
    answer: str
    model_used: str
    fallback_used: bool = False
    tool_calls: list[AgentToolCallView]
    data: Any | None = None
    visualization: AgentVisualization | None = None


class ConversationSummary(BaseModel):
    id: int
    title: str
    message_count: int
    created_at: datetime
    updated_at: datetime


class AgentMessageView(BaseModel):
    id: int
    role: Literal["user", "assistant"]
    content: str
    model_used: str | None
    tool_calls: list[AgentToolCallView]
    created_at: datetime
